File: Codes_morgan/TOURE/Pb_inverse/2.Illustrations-Python-et-Matlab/2.Illustrations-Python-et-Matlab/1._Conditionnement/Tests_cond.py
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
#from math import *
from numpy import *
#from scipy import * 
#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
print("",end='\n\n\n') 
print("Tests de conditionnements", end='\n\n')

#%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
print("----------- Matrice de Wilson --------------", end='\n\n') ;  
A = array ([[10, 7, 8, 7], [7 ,5, 6, 5], [8, 6, 10, 9], [7, 5, 9, 10]])
print(A)
#print(A.shape)
input() 

detA = linalg.det(A) 
print("Déterminant de A = ", detA, end='\n\n')
input() 

A_inv = linalg.inv(A)
print("Inverse de A", end='\n\n')
print(A_inv, end='\n\n')
input() 

cond1 = linalg.cond(A,1)
condinf = linalg.cond(A,inf)
cond2 = linalg.cond(A)
print("Conditionnement en norme 2 : ", cond2, end='\n\n')
input() 

vp,Vp = linalg.eig(A)
print("Valeurs propres : ", vp, end='\n\n')
input() 

b = array([[32], [23], [33], [31]])
print("Donnée =", transpose(b), end='\n\n')
#print(b.shape)
input() 

x = dot(linalg.inv(A),b)
print("Solution de Ax=b : x = ", transpose(x))
input() 

b_err = b + array([[0.1], [-0.1], [0.1], [-0.1]])
print("Donnée perturbée =", transpose(b_err), end='\n\n')
erreur_donnee = linalg.norm(b-b_err)/linalg.norm(b)*100 ; 
print("Erreur sur la donnée b + db = {: .2f} % ".format(erreur_donnee), end='\n\n') ; 
input() 

x_err=linalg.solve(A, b_err)
print("Solution de Ax=b perturbé : x + dx = ", transpose(x_err), end='\n\n')
erreur_sol = linalg.norm(x-x_err)/linalg.norm(x)*100 
print("Erreur sur la solution = {: .2f} %".format(erreur_sol), end='\n\n') ; 
input() 

input("...") ; print("",end='\n\n') 


# %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

print("------------ Matrices de Hilbert -----------", end='\n\n')

for n in range(1,21):
    b = zeros(n)  
    H = zeros((n,n))
    for i in range(0,n):
        S = 0 
        for j in range(0,n):
            H[i,j] = 1/(i+j+1) 
            S = S + H[i,j] 
        b[i] = S
    cond2 = linalg.cond(H) 
    print("Le conditionnement en norme 2 pour n = {} vaut :  {: .2e}".format(n,cond2)) 
    x = linalg.solve(H, b)
    print("La solution de Hx = b est x = ", transpose(x))
    input()

File: Codes_morgan/TOURE/Pb_inverse/2.Illustrations-Python-et-Matlab/2.Illustrations-Python-et-Matlab/Old/1.Conditionnement/Tests_cond.m

clear all ; close all ; 

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
fprintf('\n\n\n') ;  
fprintf('\n *********** Matrice de Wilson ************\n') ;

A = [ 10 7 8  7 ;
      7  5 6  5 ;
      8 6  10 9 ;
      7 5  9 10]

  
pause
invA = inv(A) 

pause
detA = det(A) 

pause
cond1 = cond(A,1)
condinf = cond(A,inf)
cond2 = cond(A)

pause
vp = eig(A)

pause
b = [32,23,33,31]'
x = A\b

pause
b_err = b + [0.1,-0.1,0.1,-0.1]'
erreur_donnee = norm(b-b_err)/norm(b)*100 ; 
fprintf('Erreur sur la donnée = %2.2f %%\n',erreur_donnee) ; 
pause
x_err = A\b_err
erreur_sol = norm(x-x_err)/norm(x)*100 ;
fprintf('Erreur sur la solution = %2.2f %%\n',erreur_sol) ; 

pause
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
fprintf('\n\n\n') ;  
fprintf('*********  Matrices de Hilbert **********\n') ;

for n = 2:30
   b = zeros(n,1) ; 
   for i = 1:n
   S = 0 ; 
       for j=1:n
           H(i,j) = 1/(i+j-1) ; 
           S = S + H(i,j) ; 
       end
   b(i) = S ; 
   end
   % H
   
   cond2 = cond(H) 
    pause
    x = H\b
    pause
end
%plot(1:9,condv) ; 



%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


% 
% %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% fprintf('\n **********  Matrices aleatoires  ***********\n') ;
% 
% %condv = zeros(9,1) ; 
% n = 512 ; 
% for nexp = 1:1000 
%    B = rand(n,n)-1/2*ones(n,n) ; % la grandeur peut etre ramenee à [0,1]... 
%    cond2 = cond(B) 
%    cond2bis = cond(B+7*eye(n,n))
%    pause
% end
% %plot(1:9,condv) ; 
% 
% 

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%%%%%%%% Matrice de derivation
% n = 1000 ;
% 
% u = -1*ones(1,n-1) ;
% v = ones(1,n) ;
% InvA = diag(u,-1) + diag(v) ; 
% A = inv(InvA) ; 
% 
% cond(A,1) 
% 
% eps = 0.01 ; 
% 
% Atp = inv(A'*A + eps*eye(n))*A' ;
% 
% cond(Atp,1) 